### PR TITLE
fix: fix wrong 503 Service Unavailable on pg error 53400

### LINKED ISSUE #3257: Wrong `503 Service Unavailable` on pg `53400`
## Problem

Having:

```sql
create or replace function temp_file_limit()
returns bigint as $$
  set temp_file_limit to '1MB';
  select COUNT(*) FROM generate_series('-infinity'::TIMESTAMP, 'epoch'::TIMESTAMP, INTERVAL '1 DAY');
$$ language sql security definer;
# got the snippet from https://www.postgresql.org/message-id/flat/CADkLM%3DdDM4TjiiBop4JWKz0jGk-XRX1nAovQd5cmzTODNbCr-g%40mail.gmail.com#17d18bf6206d36b74ff6404661cdc0a8
```

Generates:

```bash
curl localhost:3000/rpc/temp_file_limit -i 
HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Date: Fri, 23 Feb 2024 22:51:47 GMT
Server: postgrest/12.1 (160caaf)
Retry-After: 0
Content-Type: application/json; charset=utf-8

{"code":"53400","details":null,"hint":null,"message":"temporary file size exceeds temp_file_limit (1024kB)"}
```

Which is wrong as were' still up:

```
curl "localhost:3000/rpc/add_them?a=3&b=4"
7
```

## Solution

Turn `53xxx` into the 500 status.

(There's [507 Insufficient Storage](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/507) but it's not accurate in every case)

### PR DESCRIPTION
**Pull request by taimoorzaeem**
Fixes #3257.

### CODE REVIEW

File: test/spec/fixtures/schema.sql [hunk 1]
```diff
@@ -3759,3 +3759,9 @@ create aggregate test.outfunc_agg (anyelement) (
 , stype = "pg/outfunc"
 , sfunc = outfunc_trans
 );
+
+create or replace function temp_file_limit()
+returns bigint as $$
+  set temp_file_limit to '1MB';
+  select COUNT(*) FROM generate_series('-infinity'::TIMESTAMP, 'epoch'::TIMESTAMP, INTERVAL '1 DAY');
+$$ language sql security definer;
```
[the remark this answers is not in the document]
Comment (taimoorzaeem) [reply test/spec/fixtures/schema.sql:3767]: @wolfgangwalther 
```sql
grant set on parameter temp_file_limit to postgrest_test_anonymous;

create or replace function temp_file_limit()
returns bigint as $$
  select COUNT(*) FROM generate_series('-infinity'::TIMESTAMP, 'epoch'::TIMESTAMP, INTERVAL '1 DAY');
$$ language sql security definer set temp_file_limit to '1kB';
```
With the above, I am getting the following error on running this test:
```
but got:  {"code":"22023","details":null,"hint":"Valid units for this parameter are \"B\", \"kB\", \"MB\", \"GB\", 
and \"TB\".","message":"invalid value for parameter \"temp_file_limit\": \"1kb\""}
```
The test magically lowercases all parameter values. Same goes for `1MB` to `1mb` and then declaring it invalid.

[the remark this answers is not in the document]
Comment (taimoorzaeem) [reply test/spec/fixtures/schema.sql:3767]: Yes, I am sure. I am working on a different branch. Let me share the PR with you.

[the remark this answers is not in the document]
Comment (taimoorzaeem) [reply test/spec/fixtures/schema.sql]: Should I add another commit first to remove the `lower(...)` here? That way we don't have to wait for #3358 to be merged before we merge this.

[the remark this answers is not in the document]
Comment (taimoorzaeem) [reply test/spec/fixtures/schema.sql]: @wolfgangwalther @laurenceisla, Apparently, `GRANT SET` does not work for versions `9.6 - 14` so the test is failing for these versions. How do I get around the permission issue for these versions?

[the remark this answers is not in the document]
Comment (taimoorzaeem) [reply test/spec/fixtures/schema.sql]: > Can this test run with a different role, which you then make SUPERUSER for this purpose?

How do I run just a single test with a different role? Changing the role would by itself might require superuser privileges.

[the remark this answers is not in the document]
Comment (taimoorzaeem) [reply test/spec/fixtures/schema.sql]: @steve-chavez Any idea on this?

[the remark this answers is not in the document]
Comment (laurenceisla) [reply test/spec/fixtures/schema.sql]: You could create another superuser role like: 

```sql
`CREATE ROLE postgrest_test_superuser WITH SUPERUSER;`
```

Then grant usage on the schema, and do the request impersonating that user:

```haskell
    context "test function temp_file_limit" $
      let auth = authHeaderJWT "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoicG9zdGdyZXN0X3Rlc3Rfc3VwZXJ1c2VyIiwiaWQiOiJqZG9lIn0.LQ-qx0ArBnfkwQQhIHKF5cS-lzl0gnTPI8NLoPbL5Fg" in
      it "should return http status 500" $
        request methodGet "/rpc/temp_file_limit" [auth] "" `shouldRespondWith`
          [json|{"code":"53400","message":"temporary file size exceeds temp_file_limit (1kB)","details":null,"hint":null}|]
          { matchStatus = 500 }
```

You can copy paste to your test, it should work as-is. Additionally, you can do a `if pgVersion>=15000` for the tests (without impersonation) and the SQL you already have. Something similar to what is done in IO tests: https://github.com/PostgREST/postgrest/blob/9d1dc783bf3b736ca8e09aecfb2dc2fa214a9c78/test/io/fixtures.sql#L22-L27

[the remark this answers is not in the document]
Comment (laurenceisla) [reply test/spec/fixtures/schema.sql]: That's a different error, though. You need to grant that role to the authenticator here:

https://github.com/PostgREST/postgrest/blob/0d13b842a6c3a8d5eeb1ea7aa750d44ecff63044/test/spec/fixtures/roles.sql#L6

File: src/PostgREST/Error.hs [hunk 2]
```diff
@@ -474,7 +474,7 @@ pgErrorStatus authed (SQL.SessionUsageError (SQL.QueryError _ _ (SQL.ResultError
         '3':'9':_ -> HTTP.status500 -- external routine invocation
         '3':'B':_ -> HTTP.status500 -- savepoint exception
         '4':'0':_ -> HTTP.status500 -- tx rollback
-        '5':'3':_ -> HTTP.status503 -- insufficient resources
+        '5':'3':_ -> HTTP.status500 -- insufficient resources
```
Comment (steve-chavez) [inline src/PostgREST/Error.hs:477]: How about matching `53400` exactly here for status500 and then leaving the other `53xxx` as 503?

As mentioned on https://github.com/PostgREST/postgrest/issues/3257#issuecomment-1962327567, the other `53xxx` do seem fit for status503.

**Comment by steve-chavez** [comment]
@taimoorzaeem This also requires an update to docs https://postgrest.org/en/v12/references/errors.html#http-status-codes

File: test/spec/Feature/Query/RpcSpec.hs [hunk 3]
```diff
@@ -1478,3 +1478,10 @@ spec actualPgVersion =
             "details":"DETAIL is missing in the RAISE statement",
             "hint":"DETAIL must be a JSON object with obligatory keys: 'status', 'headers' and optional key: 'status_text'."}|]
           { matchStatus = 500 }
+
+    context "test function temp_file_limit" $
+      let auth = authHeaderJWT "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoicG9zdGdyZXN0X3Rlc3Rfc3VwZXJ1c2VyIiwiaWQiOiJqZG9lIn0.LQ-qx0ArBnfkwQQhIHKF5cS-lzl0gnTPI8NLoPbL5Fg" in
```
Comment (steve-chavez) [inline test/spec/Feature/Query/RpcSpec.hs:1483]: Hm, here you're using the `postgrest_test_superuser` right? Maybe add a comment so it's clearer?

---

Ideally we would have a function on the haskell tests, like this one on the io tests:

https://github.com/PostgREST/postgrest/blob/85a4e35a2de869ddf9b6c5f6893a3154767cbc76/test/io/test_io.py#L45

So we don't have to paste the JWT.. this could be another PR.
